File: android/app/src/main/java/com/example/mycarnavi/PlacesApiClient.kt
package com.example.mycarnavi

import android.util.Log
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

data class ParkingPlace(
    val name: String,
    val latitude: Double,
    val longitude: Double
)

class PlacesApiClient(private val apiKey: String) {

    private val client = OkHttpClient()

    /** ブロッキング呼び出しのため、IOスレッドから呼ぶこと */
    fun searchParking(
        centerLatitude: Double,
        centerLongitude: Double,
        radiusMeters: Double = 1000.0
    ): List<ParkingPlace> {

        val requestJson = JSONObject().apply {
            put("includedTypes", JSONArray().put("parking"))
            put("maxResultCount", 20)
            put("locationRestriction", JSONObject().put(
                "circle", JSONObject()
                    .put("center", JSONObject()
                        .put("latitude", centerLatitude)
                        .put("longitude", centerLongitude))
                    .put("radius", radiusMeters)
            ))
        }

        val request = Request.Builder()
            .url("https://places.googleapis.com/v1/places:searchNearby")
            .addHeader("X-Goog-Api-Key", apiKey)
            .addHeader(
                "X-Goog-FieldMask",
                "places.displayName,places.location"
            )
            .post(
                requestJson.toString()
                    .toRequestBody("application/json".toMediaType())
            )
            .build()

        client.newCall(request).execute().use { response ->

            val body = response.body?.string()

            if (!response.isSuccessful || body == null) {
                Log.e(TAG, "Places API error: code=${response.code} body=$body")
                return emptyList()
            }

            val places = JSONObject(body).optJSONArray("places")
                ?: return emptyList()

            return (0 until places.length()).mapNotNull { i ->
                val place = places.getJSONObject(i)
                val location = place.optJSONObject("location")
                    ?: return@mapNotNull null

                val lat = location.optDouble("latitude", Double.NaN)
                val lng = location.optDouble("longitude", Double.NaN)
                if (lat.isNaN() || lng.isNaN()) return@mapNotNull null

                ParkingPlace(
                    name = place.optJSONObject("displayName")
                        ?.optString("text")
                        ?: "駐車場",
                    latitude = lat,
                    longitude = lng
                )
            }
        }
    }

    companion object {
        private const val TAG = "PlacesApiClient"
    }
}

File: ios/Navi/Services/PlacesApiClient.swift
import Foundation
import OSLog

struct PlacesApiClient {
    private let apiKey: String
    private let logger = Logger(subsystem: "suzukiken.Navi", category: "PlacesApiClient")

    init(apiKey: String = MapsApiConfig.apiKey) {
        self.apiKey = apiKey
    }

    func searchParking(
        centerLatitude: Double,
        centerLongitude: Double,
        radiusMeters: Double = 1000
    ) async -> [ParkingPlace] {
        guard !apiKey.isEmpty else {
            logger.error("MAPS_API_KEY is not configured")
            return []
        }

        let body: [String: Any] = [
            "includedTypes": ["parking"],
            "maxResultCount": 20,
            "locationRestriction": [
                "circle": [
                    "center": [
                        "latitude": centerLatitude,
                        "longitude": centerLongitude,
                    ],
                    "radius": radiusMeters,
                ],
            ],
        ]

        guard let url = URL(string: "https://places.googleapis.com/v1/places:searchNearby"),
              let jsonData = try? JSONSerialization.data(withJSONObject: body)
        else {
            return []
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = jsonData
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-Goog-Api-Key")
        request.setValue(
            "places.displayName,places.location",
            forHTTPHeaderField: "X-Goog-FieldMask"
        )

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                logger.error("Places API error: \((response as? HTTPURLResponse)?.statusCode ?? -1)")
                return []
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let places = json["places"] as? [[String: Any]]
            else {
                return []
            }

            return places.compactMap { place in
                guard let location = place["location"] as? [String: Any],
                      let lat = location["latitude"] as? Double,
                      let lng = location["longitude"] as? Double
                else {
                    return nil
                }

                let name = (place["displayName"] as? [String: Any])?["text"] as? String ?? "駐車場"
                return ParkingPlace(name: name, latitude: lat, longitude: lng)
            }
        } catch {
            logger.error("Places API request failed: \(error.localizedDescription)")
            return []
        }
    }
}
